### PR TITLE
Axum integration

### LINKED ISSUE #986: axum exemple with websockets
**Is your feature request related to a problem? Please describe.**

I'm trying to build a small graphql server on top of [Axum](https://github.com/tokio-rs/axum) but can't figure out how to get websockets working with it

**Describe the solution you'd like**

I'd like an example similar to https://github.com/graphql-rust/juniper/tree/master/examples/warp_subscriptions

**Describe alternatives you've considered**

async-graphql does have an [Axum integration](https://github.com/async-graphql/async-graphql/tree/master/integrations/axum) but it does not support implementing dynamic schemas (and they don't plan to)

### PR DESCRIPTION
**Pull request by btielen**
This PR provides an integration for Axum including subscriptions (fixes #986)

- including http tests run by `juniper::http::tests::run_http_test_suite`
- including websocket tests run by `juniper::http::tests::run_ws_test_suite`

## Get started

Run `cargo run -p juniper_axum --example simple` from the root directory, and navigate to `127.0.0.1:3000` in your browser.

## Other changes
Minor changes were made in other libraries

- Added an Error for the implementation of `Sink` for `juniper_graphql_ws::Connection`, so the code doesn't panic.
- Added an extra test in `juniper::http::tests::run_ws_test_suite` to test POST requests with variables

## Discussion

- Documentation probably needs some polishing
- `juniper::http::GraphQLRequest` now has a property variables with type `Option<InputValue<S>>`, I think (but not sure) things would be easier if this has the type `Variables<S>`
- I didn't test requests with an `operation_name` defined

Any help polishing this PR is appreciated :)

### CODE REVIEW

**Comment by btielen** [comment]
In case this PR will get merged, it would be nice to update the projects documentation and book to point out an integration for Axum exists.

File: juniper_axum/src/subscriptions.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    type Error = serde_json::Error;
+
+    fn try_from(msg: AxumMessage) -> serde_json::Result<Self> {
+        serde_json::from_slice(&msg.0.into_data())
+    }
+}
+
+/// Redirect the axum [`Websocket`] to a juniper [`Connection`] and vice versa.
+///
+/// # Example
```
Comment (btielen) [inline juniper_axum/src/subscriptions.rs:28]: Provide example

File: juniper_axum/tests/juniper_http_test_suite.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    fn get(&self, url: &str) -> TestResponse {
+        let request = Request::get(url).body(Body::empty()).unwrap();
+
+        self.make_request(request)
+    }
+
+    fn post_json(&self, url: &str, body: &str) -> TestResponse {
+        let request = Request::post(url)
+            .header("content-type", "application/json")
+            .body(Body::from(body.to_string()))
```
Comment (btielen) [inline juniper_axum/tests/juniper_http_test_suite.rs:59]: Is Body::from(body) possible?

File: juniper_axum/tests/juniper_http_test_suite.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            .unwrap();
+
+        self.make_request(request)
+    }
+
+    fn post_graphql(&self, url: &str, body: &str) -> TestResponse {
+        let request = Request::post(url)
+            .header("content-type", "application/graphql")
+            .body(Body::from(body.to_string()))
```
Comment (btielen) [inline juniper_axum/tests/juniper_http_test_suite.rs:68]: Is Body::from(body) possible?

File: juniper_axum/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+homepage = "https://github.com/graphql-rust/juniper/tree/master/juniper_axum"
+repository = "https://github.com/graphql-rust/juniper"
+readme = "README.md"
+categories = ["asynchronous", "web-programming", "web-programming::http-server"]
+keywords = ["graphql", "juniper", "axum", "websocket"]
+exclude = ["/release.toml"]
+
+[dependencies]
+axum = { version = "0.5.11", features = ["ws"]}
+juniper = { path = "../juniper" }
```
Comment (btielen) [inline juniper_axum/Cargo.toml:19]: Is a version required?

**Review by btielen — COMMENTED** [review]
Some notes

File: juniper_axum/Cargo.toml (hunk 4 above)
Comment (tyranron) [reply juniper_axum/Cargo.toml:19]: @btielen in the same way as other integrations have. The version bump is automated with `cargo-release`. See `release.toml` files for examples.

File: juniper_axum/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+========================
+
+All user visible changes to `juniper_axum` crate will be documented in this file. This project uses [Semantic Versioning 2.0.0].
+
+
+
+
+## master
+
+### BC Breaks
```
Comment (LegNeato) [inline juniper_axum/CHANGELOG.md:11]: Just remove this section and put something like:

* Initial release

File: juniper_axum/CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+- Switched to 0.16 version of [`juniper` crate].
+
+
+
+
+## Previous releases
+
+See [old CHANGELOG](/../../blob/juniper_warp-v0.0.0/juniper_axum/CHANGELOG.md).
```
Comment (LegNeato) [inline juniper_axum/CHANGELOG.md:20]: This mentions `warp` and should be deleted.

File: juniper_axum/Cargo.toml (hunk 4 above)
Comment (LegNeato) [inline juniper_axum/Cargo.toml:19]: As mentioned in the PR comments already, this should have a version and wired up to the release machinery.

File: juniper_axum/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+serde = "1.0"
+serde_json = "1.0"
+juniper_graphql_ws = { path="../juniper_graphql_ws" }
```
Comment (LegNeato) [inline juniper_axum/Cargo.toml:22]: Ditto.

File: juniper_axum/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
+`juniper_axum` crate
+====================
+
+[![Crates.io](https://img.shields.io/crates/v/juniper_axum.svg?maxAge=2592000)](https://crates.io/crates/juniper_warp)
+[![Documentation](https://docs.rs/juniper_warp/badge.svg)](https://docs.rs/juniper_warp)
```
Comment (LegNeato) [inline juniper_axum/README.md:5]:
```suggestion
[![Documentation](https://docs.rs/juniper_axum/badge.svg)](https://docs.rs/juniper_axum)
```

File: juniper_axum/README.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+
+[`juniper`]: https://docs.rs/juniper
+[`juniper_axum`]: https://docs.rs/juniper_axum
+[`axum`]: https://docs.rs/axum
+[GraphQL]: http://graphql.org
+[Juniper Book]: https://graphql-rust.github.io
+[Rust]: https://www.rust-lang.org
+
+[1]: https://github.com/graphql-rust/juniper/blob/master/juniper_warp/examples/warp_server.rs
```
Comment (LegNeato) [inline juniper_axum/README.md:93]: Fix URL.

File: juniper_warp/src/lib.rs [hunk 10]
```diff
@@ -376,14 +377,14 @@ pub mod subscriptions {
         /// Errors that can happen while serializing outgoing messages. Note that errors that occur
         /// while deserializing incoming messages are handled internally by the protocol.
         Serde(serde_json::Error),
+
+        /// Errors that can happen while communication with Juniper
```
Comment (LegNeato) [inline juniper_warp/src/lib.rs:381]:
```suggestion
        /// Errors that can happen while communicating with Juniper
```

**Review by LegNeato — CHANGES_REQUESTED** [review]
Wow, this is an **AMAZING PR**  🥳 . Thank you for the change, adding docs, and adding tests! I especially like that you improved the `ws` stuff while you where here.

Some small nits but it looks solid otherwise.

File: juniper_axum/examples/simple.rs [hunk 11]
```diff
@@ -29,34 +31,34 @@ impl Query {
...
+
 #[tokio::main]
 async fn main() {
-    let schema = Arc::new(AppSchema::new(Query, EmptyMutation::new(), Subscription));
+    let schema = AppSchema::new(Query, EmptyMutation::new(), Subscription);
 
     let context = Context;
 
     let app = Router::new()
-        .route("/", get(|| playground("/graphql", Some("/subscriptions"))))
```
Comment (ilslv) [inline juniper_axum/examples/simple.rs]: @btielen Great job, thanks! I've made some insignificant changes though. Most notably simplifying the usage of the `playground`, now it feels to me a lot more `axum`y.

File: juniper_axum/examples/simple.rs [hunk 12]
```diff
@@ -72,20 +74,20 @@ async fn main() {
 }
 
 pub async fn juniper_subscriptions(
-    Extension(schema): Extension<Arc<AppSchema>>,
```
Comment (ilslv) [inline juniper_axum/examples/simple.rs]: Also I've removed a need to wrap all `Scheme`s into `Arc`s.

**Review by ilslv — APPROVED** [review]

**Comment by btielen** [comment]
Thanks for the feedback, I will take a look at it tomorrow.

**Comment by harksin** [comment]
:wave: Incredible work :star_struck:  , Is there any plan to merge & release it ?

**Comment by urothis** [comment]
Any blockers on merging this?

**Comment by Cugatay** [comment]
Hey, are you going to push Axum integration? I loved Axum when I tried it, and I want to use something I already know when using GraphQL

**Comment by LegNeato** [comment]
There are still outstanding review comments here...

**Comment by tyranron** [comment]
I plan to finish it in next few weeks.

**Comment by tyranron** [comment]
#### Progress

1. Reverted `graphql_ws::WebsocketError` changes, as they're clearly behind theme of this PR.
2. The same with `Clone`ing of `RootNode`. Such refactoring is better to be done properly in a separate PR (if makes sense), as touches all the integration crates.
3. Added CI integration for `juniper_axum`.
4. Added GraphiQL endpoint to `juniper_axum`.
5. Migrated to 0.6 `axum` (without subscriptions yet).
6. Got rid of `GetQueryVariables` and `JsonRequestBody`, which duplicated `GraphQLRequest`.

**Comment by tyranron** [comment]
#### Progress

1. Parametrized `extractor::JuniperRequest` with `ScalarValue`.
2. Reworked legacy `graphql-ws` GraphQL over WebSocket Protocol implementation.
3. Fixed doc tests.
4. Added new `graphql-transport-ws` GraphQL over WebSocket Protocol implementation.
5. Implemented auto-selection between new `graphql-transport-ws`  and legacy `graphql-ws` GraphQL over WebSocket Protocol implementations.
6. Add ready-to-go default `graphql`, `subscriptions::ws`, `subscriptions::graphql_ws` and `subscriptions::graphql_trasnport_ws` handlers.
7. Polished API docs.

**Comment by tyranron** [comment]
@LegNeato I'm going to merge this in next few days. Take a final look, if you would like. All the stuff is ready, only minor CI issues are left to be resolved.

**Review by tyranron — APPROVED** [review]

**Pull request merged**